pkg/controller/identity/iamrole/controller.go

```diff
@@ -89,78 +91,77 @@ type external struct {
 }
 
 func (e *external) Observe(ctx context.Context, mgd resource.Managed) (managed.ExternalObservation, error) {
-	cr, ok := mgd.(*v1alpha3.IAMRole)
+	cr, ok := mgd.(*v1beta1.IAMRole)
 	if !ok {
 		return managed.ExternalObservation{}, errors.New(errUnexpectedObject)
 	}
 
 	req := e.client.GetRoleRequest(&awsiam.GetRoleInput{
-		RoleName: aws.String(cr.Spec.RoleName),
+		RoleName: aws.String(meta.GetExternalName(cr)),
 	})
 
 	observed, err := req.Send(ctx)
 
-	if iam.IsErrorNotFound(err) {
-		return managed.ExternalObservation{
-			ResourceExists: false,
-		}, nil
-	}
-
 	if err != nil {
-		return managed.ExternalObservation{}, errors.Wrapf(err, errGet, cr.Spec.RoleName)
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(iam.IsErrorNotFound, err), errGet)
 	}
 
 	cr.SetConditions(runtimev1alpha1.Available())
 
-	cr.UpdateExternalStatus(*observed.Role)
+	cr.Status.AtProvider = iam.GenerateRoleObservation(*observed.Role)
 
 	return managed.ExternalObservation{
 		ResourceExists: true,
 	}, nil
 }
 
 func (e *external) Create(ctx context.Context, mgd resource.Managed) (managed.ExternalCreation, error) {
-	cr, ok := mgd.(*v1alpha3.IAMRole)
+	cr, ok := mgd.(*v1beta1.IAMRole)
 	if !ok {
 		return managed.ExternalCreation{}, errors.New(errUnexpectedObject)
 	}
 
 	cr.Status.SetConditions(runtimev1alpha1.Creating())
 
-	req := e.client.CreateRoleRequest(&awsiam.CreateRoleInput{
-		RoleName:                 aws.String(cr.Spec.RoleName),
-		AssumeRolePolicyDocument: aws.String(cr.Spec.AssumeRolePolicyDocument),
-		Description:              aws.String(cr.Spec.Description),
-	})
+	name := meta.GetExternalName(cr)
+	if name == "" {
+		name = cr.GetGenerateName()
+	}
+
+	req := e.client.CreateRoleRequest(iam.GenerateCreateRoleInput(name, &cr.Spec.ForProvider))
 
-	result, err := req.Send(ctx)
+	_, err := req.Send(ctx)
 	if err != nil {
 		return managed.ExternalCreation{}, errors.Wrap(err, errCreate)
 	}
 
-	cr.UpdateExternalStatus(*result.Role)
-
 	return managed.ExternalCreation{}, nil
 }
 
 func (e *external) Update(ctx context.Context, mgd resource.Managed) (managed.ExternalUpdate, error) {
```

Comment:
The `Update` methods for `IAMRole` and `IAMRolePolicyAttachment` will never be called because we are not checking if `IsUpToDate` in `Observe` method. This can be a tricky thing to get right, but here is a good reference: 

https://github.com/crossplane/stack-gcp/blob/8ece46f91e44e4139ba92f848be687b970acf214/pkg/controller/database/cloudsql.go#L150

https://github.com/crossplane/stack-gcp/blob/8ece46f91e44e4139ba92f848be687b970acf214/pkg/controller/database/cloudsql.go#L156

Let me know if you need more detail :)